chain/chain/src/chain.rs

```diff
@@ -2407,6 +2414,84 @@ impl Chain {
         )
     }
 
+    fn get_epoch_start_sync_hash_impl(
+        &self,
+        sync_header: &BlockHeader,
+    ) -> Result<CryptoHash, Error> {
+        let mut epoch_id = *sync_header.epoch_id();
+        let mut hash = *sync_header.hash();
+        let mut prev_hash = *sync_header.prev_hash();
+        loop {
+            if prev_hash == CryptoHash::default() {
+                return Ok(hash);
+            }
+            let header = self.get_block_header(&prev_hash)?;
+            if &epoch_id != header.epoch_id() {
+                return Ok(hash);
+            }
+            epoch_id = *header.epoch_id();
+            hash = *header.hash();
+            prev_hash = *header.prev_hash();
+        }
+    }
+
+    /// Find the hash of the first block on the same epoch (and chain) of block with hash `sync_hash`.
+    pub fn get_epoch_start_sync_hash(&self, sync_hash: &CryptoHash) -> Result<CryptoHash, Error> {
+        let header = self.get_block_header(sync_hash)?;
+        self.get_epoch_start_sync_hash_impl(&header)
+    }
+
+    // Returns the first block for which at least two new chunks have been produced for every shard in the epoch
+    pub fn get_current_epoch_sync_hash(
```

Comment:
Hmm, yea like I said above I moved get_epoch_start_sync_hash() out of the state sync code, but maybe we could move them both back over there? The reason I moved it to chain.rs is that previously, chain.rs was not referencing the state sync code to [decide whether it wants to make a new snapshot](https://github.com/near/nearcore/blob/70acea66bac9e5eeee64c0f417112f762e2a2b55/chain/chain/src/chain.rs#L3922). To do that, it could just check if the block was the first of the epoch. But really what we’re checking is whether the next block will be the “sync_hash” block. It’s just that those two happened to be the same before. But now that there’s some more logic involved in deciding that, I moved it here to not make chain.rs depend on any state sync files. But actually I think after I do the FIXME in here (that says we should just not call this inefficient code in this spot anyway), I think we’ll be able to move get_epoch_start_sync_hash() back to where it was before, because here we won’t need to call that in deciding if the next block is the sync block

Reply:
Just to be clear, I'm not suggesting to do it in this PR. Just generally state sync could use some refactoring. If you see any potential improvements like those it would be awesome to keep track of it and implement some.